post_office/admin.py

```diff
@@ -64,16 +90,113 @@ class EmailAdmin(admin.ModelAdmin):
+            if template is None:
+                # cache compiled template to speed up rendering of list view
+                template = Template(instance.template.subject)
+                setattr(self, template_cache_key, template)
+            subject = template.render(Context(instance.context))
+        else:
+            subject = instance.subject
+        return Truncator(subject).chars(100)
+
+    subject_shortened.short_description = _("Subject")
+    subject_shortened.admin_order_field = 'subject'
+
+    def use_template(self, instance):
+        return bool(instance.template)
+
+    use_template.short_description = _("Use Template")
+    use_template.boolean = True
+
+    def get_fieldsets(self, request, obj=None):
+        fieldsets = [
+            (None, {
+                'fields': ['from_email', 'to', 'cc', 'bcc',
+                           'priority', ('status', 'scheduled_time')],
+            }),
+        ]
+        email_plain, email_html = False, False
+        for message in obj.email_message().message().walk():
+            if not isinstance(message, SafeMIMEText):
+                continue
+            content_type = message.get_content_type()
+            if content_type == 'text/plain':
+                email_plain = True
+            elif content_type == 'text/html':
+                email_html = True
+
+        if email_html:
+            fieldsets.append(
+                (_("HTML Email"), {'fields': ['subject_rendered', 'render_body_html']})
+            )
+            if email_plain:
+                fieldsets.append(
+                    (_("Text Email"), {'classes': ['collapse'], 'fields': ['render_body_plain']})
+                )
+        elif email_plain:
+            fieldsets.append(
+                (_("Text Email"), {'fields': ['subject_rendered', 'render_body_plain']})
+             )
+
+        return fieldsets
+
+    def subject_rendered(self, instance):
+        message = instance.email_message()
+        return format_html('<strong>{}</strong>', message.subject)
+
+    subject_rendered.short_description = _("Subject")
+
+    def render_body_plain(self, instance):
+        for message in instance.email_message().message().walk():
+            if isinstance(message, SafeMIMEText) and message.get_content_type() == 'text/plain':
+                return format_html('<pre>{}</pre>', message.get_payload())
+
+    render_body_plain.short_description = _("Mail Body")
+
+    def render_body_html(self, instance):
+        pattern = re.compile('cid:([0-9a-f]{32})')
+        url = reverse('admin:post_office_email_image', kwargs={'pk': instance.id, 'content_id': 32 * '0'})
+        url = url.replace(32 * '0', r'\1')
+        for message in instance.email_message().message().walk():
+            if isinstance(message, SafeMIMEText) and message.get_content_type() == 'text/html':
+                payload = message.get_payload(decode=True).decode('utf-8')
+                body = mark_safe(pattern.sub(url, payload))
```

Comment:
I'm worried that this could cause a stored XSS attack. Does Django have a built in way to mark template strings as safe, except the JS part?

Reply:
Good point, this indeed is a possible attack vector. While email clients usually do not execute JavaScript, by adding a `<script>doSomethingEvil();</script>` to the body, it would be executed inside the Django admin.

Django itself does not provide any tools to remove `<script>`-tags. Instead they recommend to use the [bleach](https://pypi.org/project/bleach/) library. I could use this libary, but this would add another third party dependency. So you have to evaluate, if that additional requirement is worthwhile.

I then would also remove the `<style>`-tag. Emails clients ignore it anyway and you can do other nasty stuff with it.

Reply:
Well, one thing that's for sure is that we don't want to ship insecure apps.

A good compromise would be to only render the HTML preview if `bleach` is installed. What do you think?

Reply:
Instead of bleach, I added the lxml library as an optional dependency. Whenever it is installed, HTML emails are sanitized. Bleach was too strict and required to much special configuration.